### ISSUE TITLE
RFE: Don't snapshot view on base table snapshot

### ISSUE THREAD

**Issue by Michal-Leszczynski** [scylladb/scylladb#20760]
IIUC, calling Scylla snapshot API (`POST /storage_service/snapshots`) on only the base table results in creating snapshots for the base table and all of its views. So even if I don't care for snapshots of view tables (e.g. because they [shouldn't be used for restore purposes](https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/backup-restore/restore.html#repeat-the-following-steps-for-each-node-in-the-cluster)), they will always be created and will consume space.

Why is that a case? Shouldn't only specified tables be snapshot-ed?
I find it confusing and more difficult to maintain, as user/managing software might forget about this behavior and let the useless snapshot of view table sit on the node and consume space for a long time.
E.g. right now ScyllaDB Manager does not back up view tables sstables, but it still creates their snapshots (unknowingly) which are stored on the nodes until the end of the backup process.

Of course, view table snapshots can be deleted manually with Scylla API (`POST /storage_service/snapshots`), but one needs to remember about that.

**Comment by Michal-Leszczynski** [comment]
cc: @nyh @tchaikov

**Comment by mykaul** [comment]
CC @piodul , @bhalevy 

I wouldn't change the API to do something different, but extend it (&views=false or something?)

**Comment by nyh** [comment]
I believe this was changed deliberately by @bhalevy two years ago, in commit d32c497cd94d757711f2006926b2278957584f9a (pull request https://github.com/scylladb/scylladb/pull/11616), aimed to fix https://github.com/scylladb/scylladb/issues/11612.

We need to discuss this (we should have done it two years ago...).

**Comment by bhalevy** [comment]
Why don't we want to restore views and indices?

**Comment by Michal-Leszczynski** [comment]
> Why don't we want to restore views and indices?

We want to, but view sstables are not used for this purpose.
Let's say we have the whole schema restored in the cluster - both base table and the view table are there, but empty.
The recommended restore procedure looks like that:
- save view create stmt
- drop view
- restore base table
- create view

That's why even though Scylla Manager supports view restoration, it does not want to back up view sstables.

**Comment by piodul** [comment]
> Why don't we want to restore views and indices?

Obviously, we want to let users restore views and indices, but not by restoring their sstables but rather rebuilding them from scratch. There is no way to take a consistent snapshot of both the base table and the views. By rebuilding you avoid this problem and also make the snapshot size smaller.

**Comment by bhalevy** [comment]
> > Why don't we want to restore views and indices?
> 
> Obviously, we want to let users restore views and indices, but not by restoring their sstables but rather rebuilding them from scratch. There is no way to take a consistent snapshot of both the base table and the views. By rebuilding you avoid this problem and also make the snapshot size smaller.

Bear in mind that even the snapshot of the base table isn't atomic i.e. it isn't consistent across nodes.
For fast, 1-to-1 restore, rebuilding the views may take a long time, breaching the restore SLA "contract".
@tzach is it a requirement for 1-to-1 restore to restore the view contents from backup rather than regenerating it from scratch?
Cc @regevran

**Comment by nyh** [comment]
@bhalevy I'm still very surprised why we even went in this direction. Looking at the issues I mentioned above, it seems that in the past you could simply snapshot a table and a view separately - if the manager wanted it could snapshot both of them, and if it wanted, it could snapshot just the base. But then, one day it was decided to **outlaw** snapshotting a view. Why? As soon as this was done, and people still wanted to snapshot the view, you added your patch which caused both base and views to be snapshot together. So we lost the ability to snapshot only the base.

Wouldn't it make sense to undo all these patches and allow the base and each view to be snapshotted independently, if the user wants to do it?

**Comment by bhalevy** [comment]
> @bhalevy I'm still very surprised why we even went in this direction. Looking at the issues I mentioned above, it seems that in the past you could simply snapshot a table and a view separately - if the manager wanted it could snapshot both of them, and if it wanted, it could snapshot just the base. But then, one day it was decided to **outlaw** snapshotting a view. Why? As soon as this was done, and people still wanted to snapshot the view, you added your patch which caused both base and views to be snapshot together. So we lost the ability to snapshot only the base.
> 
> Wouldn't it make sense to undo all these patches and allow the base and each view to be snapshotted independently, if the user wants to do it?

That's doable of course, and I don't object changing the implementation.
But it is an API change that might effect existing users, so we may want to add an option for it that will retain the existing behavior on upgraded clusters and will enable the revised behavior in new clusters.

**Comment by nyh** [comment]
By the way, @bhalevy do you know what happens in the case of **auto-snapshot** and deletion of a base table with views? Do we also save both the base table and its views as snapshots?

**Comment by regevran** [comment]
> CC [@piodul](https://github.com/piodul) , [@bhalevy](https://github.com/bhalevy)
> 
> I wouldn't change the API to do something different, but extend it (&views=false or something?)

Seems reasonable.
I think we'll keep the API to mean 'both' and add an API for views only and base-only
(although I don't know what do we have to do with snapshots of views)

**Comment by nyh** [comment]
> > By the way, [@bhalevy](https://github.com/bhalevy) do you know what happens in the case of **auto-snapshot** and deletion of a base table with views? Do we also save both the base table and its views as snapshots?
> 
> We don't allow deleting the base table if it has views. Those need to be dropped in advance (not that I think it's a good thing...).

I know, but the question is what happens when you drop a view - does it save an auto-snapshot of the deleted view?
If so, is that what we want to happen? :-(

**Comment by regevran** [comment]
> If so, is that what we want to happen?

The answer depends on the answer to:
**Do we have a reason to snapshot sstable except for backup?**

If we do, then we need to look at that other reason to decide if to snapshot the dropped view.
If we don't have any other reason other than backup - then, no, no need to snapshot a dropped view.

Another question - what do we do with snapshots of dropped tables?
How do we resurrect the tables? Can we resurrect views as well?

**Comment by bhalevy** [comment]
> > If so, is that what we want to happen?
> 
> The answer depends on the answer to: **Do we have a reason to snapshot sstable except for backup?**

The motivation for auto-snapshot of truncated and dropped tables is to protect against human error if a table if truncated or dropped by mistake.
That said we don't provide a good api to use or manage those auto-snapshots, and hence they are disabled on the cloud AFAIK.

> If we do, then we need to look at that other reason to decide if to snapshot the dropped view. If we don't have any other reason other than backup - then, no, no need to snapshot a dropped view.
> 
> Another question - what do we do with snapshots of dropped tables? How do we resurrect the tables? Can we resurrect views as well?

We can resurrect dropped/truncated tables using the restore api (nodetool refresh), but we don't have any automation for that.
Views can be regenerated from scratch and it's not a good idea to restore them using nodetool refresh, so it should be safe to just not snapshot view/indices.

**Comment by nyh** [comment]
> The motivation for auto-snapshot of truncated and dropped tables is to protect against human error if a table if truncated or dropped by mistake. That said we don't provide a good api to use or manage those auto-snapshots, and hence they are disabled on the cloud AFAIK.

This is true, and the usefulness of auto-snapshot is dubious - see https://github.com/scylladb/scylladb/issues/7123 for a better discussion. But this discussion is orthogonal to the original discussion in this issue, and I'm sorry I brought it up.
